File: fizzbuzz/discord/info.py
"""Information about FizzBuzz and its run conditions."""

from datetime import UTC, datetime
from os import getpid
from pathlib import Path
from platform import platform
from sys import version_info

from discord import version_info as discord_version_info
from psutil import Process

__author__ = "Gavin Borne"
__license__ = "MIT"

_SECONDS_PER_HOUR = 3600
_SECONDS_PER_MINUTE = 60
_HOURS_PER_DAY = 24
_MINUTES_PER_HOUR = 60
_MICROSECONDS_PER_SECOND = 1_000_000

START_TIME = datetime.fromtimestamp(Process(getpid()).create_time(), tz=UTC)
"""The time the bot was started."""

HOST = platform()
"""Information about the device hosting FizzBuzz."""

PYTHON_VERSION = f"{version_info.major}.{version_info.minor}.{version_info.micro}"
"""The current Python version running FizzBuzz."""

DISCORD_DOT_PY_VERSION = (
    f"{discord_version_info.major}.{discord_version_info.minor}"
    f".{discord_version_info.micro}"
)
"""The current discord.py version running FizzBuzz."""

BOT_APP_ID = 1303870147873996902
"""The app ID of FizzBuzz."""

PROTOTYPE_BOT_APP_ID = 1437156873722921053
"""The app ID of the testing bot."""

# DON'T FORGET TO UPDATE changelog.md IF YOU'RE EDITING THIS!!
VERSION = "v2.0.0"
"""The current release version of FizzBuzz."""


def _get_dependencies() -> list[str]:
    path = Path("requirements.txt")
    if not path.is_file():
        raise FileNotFoundError(path)

    # get dependencies from requirements.txt
    dependencies: list[str] = []
    with path.open(encoding="utf-8") as file:
        dependencies.extend(line.strip() for line in file)
    return dependencies


DEPENDENCIES = _get_dependencies()
"""All the dependencies FizzBuzz relies on."""


def get_uptime() -> str:
    """Get the uptime of the bot as a string.

    Returns:
        str: The uptime of the bot.
    """
    uptime = datetime.now(UTC) - START_TIME
    days = uptime.days
    hours = (uptime.seconds // _SECONDS_PER_HOUR) % _HOURS_PER_DAY
    minutes = (uptime.seconds // _SECONDS_PER_MINUTE) % _MINUTES_PER_HOUR
    seconds = uptime.seconds % _SECONDS_PER_MINUTE
    microseconds = uptime.microseconds % _MICROSECONDS_PER_SECOND
    return (
        f"{days} days, {hours} hours, {minutes} minutes, "
        f"{seconds} seconds, {microseconds} microseconds"
    )

File: src/info.rs
//! Global information about FizzBuzz and its run conditions.

use std::sync::OnceLock;

use chrono::{DateTime, Utc};

/// The current release version of FizzBuzz.
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

/// The time that the program started.
static START_TIME: OnceLock<DateTime<Utc>> = OnceLock::new();

/// Initializes the start time of the program.
/// If the start time has already been initialized, does nothing.
pub fn init_start_time() {
    START_TIME.get_or_init(Utc::now);
}

/// Returns the start time of the program.
///
/// # Returns
///
/// - `DateTime<Utc>` - The start time.
///
/// # Panics
///
/// Panics if the start time has not been initialized.
pub fn start_time() -> DateTime<Utc> {
    *START_TIME.get().expect("Start time not initialized")
}

/// Returns the uptime of the program as a human-readable string.
///
/// # Returns
///
/// - `String` - The uptime.
pub fn get_uptime_string() -> String {
    let uptime = Utc::now() - start_time();
    let total_secs = uptime.num_seconds();
    let days = total_secs / 86400;
    let hours = (total_secs % 86400) / 3600;
    let minutes = (total_secs % 3600) / 60;
    let seconds = total_secs % 60;
    format!("{days} days, {hours} hours, {minutes} minutes, {seconds} seconds")
}

/// Get information about the hosting device.
///
/// # Returns
///
/// - `String` - The host device information.
pub fn host() -> String {
    format!(
        "{} {}",
        sysinfo::System::name().unwrap_or_else(|| "Unknown".into()),
        sysinfo::System::os_version().unwrap_or_else(|| "".into())
    )
}
